packages/hash/frontend/src/components/hooks/block-protocol-functions/knowledge/use-block-protocol-file-upload.ts

```diff
@@ -0,0 +1,170 @@
+
+// https://dev.to/qortex/compute-md5-checksum-for-a-file-in-typescript-59a4
+
+function _computeChecksumMd5(file: File): Promise<string> {
+  return new Promise((resolve, reject) => {
+    const chunkSize = 2097152; // Read in chunks of 2MB
+    const spark = new SparkMD5.ArrayBuffer();
+    const fileReader = new FileReader();
+
+    let cursor = 0;
+
+    function processChunk(chunkStart: number) {
+      const chunkEnd = Math.min(file.size, chunkStart + chunkSize);
+      fileReader.readAsArrayBuffer(file.slice(chunkStart, chunkEnd));
+    }
+
+    fileReader.onload = (evt: ProgressEvent<FileReader>) => {
+      if (!evt.target?.result) {
+        return;
+      }
+      spark.append(evt.target.result as ArrayBuffer);
+      cursor += chunkSize;
+
+      if (cursor < file.size) {
+        processChunk(cursor);
+      } else {
+        resolve(spark.end());
+      }
+    };
+
+    fileReader.onerror = () => {
+      reject(
+        new Error(
+          "An error occurred generating the md5 checksum during file upload",
+        ),
+      );
+    };
+
+    processChunk(0);
+  });
+}
+
+export const useBlockProtocolFileUpload = (
+  _readonly?: boolean,
+): { uploadFile: UploadFileRequestCallback } => {
+  const [requestFileUploadFn] = useMutation<
+    RequestFileUploadMutation,
+    RequestFileUploadMutationVariables
+  >(requestFileUpload);
+
+  const [createFileFromLinkFn] = useMutation<
+    CreateFileFromLinkMutation,
+    CreateFileFromLinkMutationVariables
+  >(createFileFromLink);
+
+  const uploadFileToStorageProvider = async (
+    presignedPostData: RequestFileUploadResponse["presignedPost"],
+    file: File,
+  ) => {
+    const formData = new FormData();
+    const { url, fields } = presignedPostData;
+
+    for (const [key, val] of Object.entries(fields)) {
+      formData.append(key, val as string);
+    }
+
+    formData.append("file", file);
+
+    return await fetch(url, {
+      method: "POST",
+      body: formData,
+    });
+  };
+
+  const uploadFile: UploadFileRequestCallback = useCallback(
+    async ({ data: fileUploadData }) => {
+      const { file, url, mediaType } = fileUploadData ?? {
+        mediaType: "image",
+      };
+      if (url?.trim()) {
+        const result = await createFileFromLinkFn({
+          variables: {
+            url,
+            mediaType,
+          },
+        });
+
+        if (!result.data) {
+          throw new Error(
+            "An error occured while creating a file from an external link",
+          );
+        }
+
+        const {
+          createFileFromLink: {
+            metadata: { editionId },
+          },
+        } = result.data;
+
+        return {
+          entityId: editionId.baseId as EntityId,
+          url,
+          mediaType,
+        };
+      }
+
+      if (!file) {
+        throw new Error(`Please provide a valid file to be uploaded`);
+      }
+
+      // const contentMd5 = await computeChecksumMd5(file);
```

Comment:
Nit: can you leave a `@todo` explaining when this will be used?

Reply:
8a5e30c22f82c83e323b91c06975d83fe382996f I ended up just removing it as I don't even know when we'd do this